### DISCUSSION TITLE
Route Aliases

### DISCUSSION THREAD

**Discussion by 7snovic** [laravel/framework#51306] [Ideas]
TL;DR
Be able to set some alias for a named route.

### Problem & Motivation
I have a Laravel project, I had to use two Laravel packages on that project. One was implementing some authentication logic `Package1` with some routes, let's say a route named `package.auth`. And the other package `Package2` was for implementing some other logic with some UI that uses a pre-defined named routes, in this case let's say `package.provider.auth`.
Typically, I need to implement some routes to be able to use the second package, the problem is that, I wanted to use the routes from the `Package1` into the second package, I know that some approaches may be like defining a new route like this:
```php
Route::get('/some-new-path', 'SomeController@method`)->name('package.provider.auth')
```

But in my case, this will cause backward incompatibility.

I went through the Laravel documentation to check how to set a route alias to the `package.provider.auth` route and set it to `package.auth`, I couldn't find anything that allows me to add some alias name to a pre-defined named route. So, I wrote a [PR](https://github.com/laravel/framework/pull/51307) to bring the functionality to Laravel routing system.

Something like:
```php
// Route::alias($alias, $namedRoute)
Route::alias('package.auth', `package.provider.auth`);
```

The mentioned [PR](https://github.com/laravel/framework/pull/51307) is still needing a lot of work to be done, something like
- Decide if the alias name is already existing as a named route, should we override it or throw some exceptions.
- Decide if a named route can have a multiple aliases
- Printing the related aliases alongside to the named routes in the `route:list` artisan command

I just wanted to check some feedbacks from the community if the functionality is worth to be implemented, or it's a kind of over load for a custom situation - my described previous situation -.

Thank you.

**Comment by shaedrich** [comment]
> Decide if a named route can have a multiple aliases

_If_ a route can have _an_ alias, does it really matter if it has one or multiple?

**Reply by 7snovic** [reply]
Not really sure, but maybe. Actually, that's why I am trying to wait for some feedback / edge-cases

**Reply by shaedrich** [reply]
Sure, you usually don't want more aliases for one route than you have routes in total (assuming, you have more than one or two routes), because you don't want this memory overhead, but for it to become a problem, it has to be a considerable amount, and in my opinion, at this point this would be the problem of the developer using it not the one implementing it.

**Reply by 7snovic** [reply]
Yes, sure, it's a valid point that it's the end-user responsibility.

**Reply by GianfriAur** [reply]
I strongly agree with this proposal. I ran into the exact same need while developing a set of modular Laravel packages, each responsible for a distinct part of an application. I ended up implementing something very close to what you described — happy to share the code and discuss a potential PR if there's interest.

My specific use case illustrates the problem well: I have a package that adds SaaS/team-based functionality, which exposes a route that must be reachable under the name `saas.team-selector`. The host application, on the other hand, defines its own entry point under the name `app`. In a particular setup, both of these turn out to point to the same route — `GET /` — which is also named `home`.

So a single route ends up needing three names simultaneously: `home`, `app`, and `saas.team-selector`. Right now, the only workarounds are either registering duplicate routes (wasteful) or scattering configuration values across packages so each one knows what to call its navigation targets — which introduces coupling and, arguably, its own overhead.

On the performance concern: is resolving an alias at routing time really more expensive than bootstrapping config files, merging them, and having each package read values out of them at runtime? I'd argue the alias approach is actually lighter, especially if the alias map is resolved once and cached alongside the route cache.

**Comment by shaedrich** [comment]
I seem to have completely forgotten about your proposal 🤦🏻‍♂️ 😅 fyi: Last year, I opened a PR (#58135) to implement route aliasing but it got rejected by Taylor

**Reply by GianfriAur** [reply]
I had already seen this PR while we were dealing with the same problem internally, but I ended up taking a quite different approach.

Rather than aliasing the route itself (URI + action), I focused exclusively on aliasing the route **name** — think of it as an alias for the `as` option. The route is registered once, as usual, and the alias is simply an alternative name that resolves to the same named route at call time (e.g. when using `route()`).

This keeps the route collection clean with no duplicates, while still allowing multiple packages to reference the same route under their own expected name.

Do you think it would be worth opening a separate PR with this approach? I'd love to get some feedback before going through the effort, and I'm also curious whether the core team would prefer one approach over the other — or a combination of both.

**Reply by shaedrich** [reply]
smart 💡

What I liked about my approach is that by allowing aliasing, we would also automatically allow extending with no additional code, however,
* in reality, most route properties are not suitable for extension
* the extendibility option might have scared off Taylor because he might have feared it to be too broad for an initial implementation

Taylor is a tough cookie but it's always worth a try 🤞🏻

**Reply by GianfriAur** [reply]
I made the RR, I hope it goes well 🤞🏻 [#59078](https://github.com/laravel/framework/pull/59078) , more than anything it's the first in the Laravel community so I hope so 😅
